### PR TITLE
Removed function value from gradient computation result.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

File: fncas/optimize.h [hunk 1]
```diff
@@ -187,7 +187,7 @@ OptimizationResult ConjugateGradientOptimizer<F>::Optimize(const std::vector<dou
   std::vector<double> current_point(starting_point);
 
   const auto initial_f_gradf = gi(current_point);
-  std::vector<double> current_grad = initial_f_gradf.gradient;
+  std::vector<double> current_grad = initial_f_gradf;
```
Comment (mzhurovich) [inline fncas/optimize.h]: Now it's not `f_gradf` but just a gradient, so I'd suggest `s/initial_f_gradf/initial_grad/`

File: fncas/optimize.h [hunk 2]
```diff
@@ -197,10 +197,10 @@ OptimizationResult ConjugateGradientOptimizer<F>::Optimize(const std::vector<dou
     const auto new_f_gradf = gi(new_point);
 
     // Calculating direction for the next step.
-    const double omega = std::max(fncas::PolakRibiere(new_f_gradf.gradient, current_grad), 0.0);
-    s = SumVectors(s, new_f_gradf.gradient, omega, -1.0);
+    const double omega = std::max(fncas::PolakRibiere(new_f_gradf, current_grad), 0.0);
```
Comment (mzhurovich) [inline fncas/optimize.h]: Same here - `new_grad`

File: fncas/differentiate.h [hunk 3]
```diff
@@ -223,12 +216,11 @@ struct g_intermediate : g {
...
-  virtual result operator()(const std::vector<fncas_value_type>& x) const {
-    result r;
-    r.value = f_(x);
-    r.gradient.resize(g_.size());
+  virtual std::vector<fncas_value_type> operator()(const std::vector<fncas_value_type>& x) const {
+    std::vector<fncas_value_type> r;
+    r.resize(g_.size());
     for (size_t i = 0; i < g_.size(); ++i) {
-      r.gradient[i] = g_[i](x, reuse_cache::reuse);
+      r[i] = g_[i](x, i ? reuse_cache::reuse : reuse_cache::invalidate);
```
Comment (mzhurovich) [inline fncas/differentiate.h:223]: :+1:

**Comment by mzhurovich** [comment]
@dkorolev LGTM, just a couple suggestions about variable names.
BTW, why do we keep `deprecated_test` in the repo? Should we do smth to make it `not_deprecated_test`? :)

**Comment by dkorolev** [comment]
@mzhurovich : Good eye on the variables, PTAL!

It would be hard to convert `deprecated_test` into a `non_deprecated_test`: The deprecated one uses auto-generated code and does not use `gtest`. At the same time, it tests arithmetical functions, so I'm glad we have those tests.

I suggest we keep it as an intern project or an interview question :-)

**Pull request merged**